### DISCUSSION TITLE
H2O formats

### DISCUSSION THREAD

**Discussion by wlandau** [ropensci/targets#466] [Ideas]
## Prework

* [x] Read and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/ropensci/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [x] New features take time and effort to create, and they take even more effort to maintain. So if the purpose of the feature is to resolve a struggle you are encountering personally, please consider first posting a "trouble" or "other" issue so we can discuss your use case and search for existing solutions first.
* [x] Format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Proposal

https://stackoverflow.com/questions/67584018/r-targets-with-h2o/67589199#67589199

**Comment by wlandau** [comment]
Looking at the `h2o` package manual, I see multiple functions for saving different kinds of objects in different kinds of storage formats:

1. `h2o.saveModel()`
1. `h2o.save_mojo()`
1. `h2o.save_frame()`
1. `h2o.saveGrid()`

On reflection, it could be a bit much to implement four different formats just for `h2o`, especially because I am not sure if they would cover all the important kinds of `h2o` objects. In contrast, `torch` was a good fit because it has a `torch_save()` function that works for all `torch` objects.

In case `targets` does not adopt `h2o`-specific formats, [file output targets](https://books.ropensci.org/targets/files.html#external-output-files) are a good workaround.

**Reply by DavZim** [reply]
Hi Will,
I have looked into the different modes `h2o` saves the data, mostly the differences between `h2o.exportFile()` and `h2o.saveModel()`. The functions just send commands to the cluster but with different configurations. 

- `h2o:::.h2o.__remoteSend(.h2o.__EXPORT_FILES(data), ...)` for [`h2o.exportFile()`](https://github.com/h2oai/h2o-3/blob/master/h2o-r/h2o-package/R/export.R#L75)
- `h2o:::.h2o.__remoteSend(paste0("Models.bin/",object@model_id), ...)` for [`h2o.saveModel()`](https://github.com/h2oai/h2o-3/blob/master/h2o-r/h2o-package/R/export.R#L249)

Furthermore, `h2o.exportFile()` takes a file argument, whereas `h2o.saveModel()` takes a path and there seems to be no way to specify the filename without renaming the file manually.
Also, there is not a single (de-) serialize function but different functions for each object type. I guess we kind of knew that, but good to check.

I have tried to find an easy, elegant, or working solution but nothing really worked without lots of edge-case-handling.

I think the workaround with manual file targets is easiest. For now I think this can be a NOFIX or leave it open, maybe someone stumbles upon a nice solution...

Thanks for your time regardless!

**Reply by wlandau** [reply]
Thanks for looking into this, @DavZim. It's good to know more definitively that this is not an easy fit right now. Maybe universal serialization would be a nice feature request for the `h2o` package, citing `keras` and `torch` as precedent.

**Comment by DavZim** [comment]
Hi Will,
thanks for the initiative around this issue and also for targets in general. It really makes the life of me and my team a lot easier! 
I am on vacation next week, but I will come back to you afterwards regarding the H2O integration!
